Base class for the quantized ConvTranspose

z-a-f opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #35460 Quantized ConvTranspose1d
* #35371 Quantized ConvTranspose2d
* **#35370 Base class for the quantized ConvTranspose**

Differential Revision: [D20641812](https://our.internmc.facebook.com/intern/diff/D20641812)

Commit:
Base class for the quantized ConvTranspose

Commit:
Update on "Base class for the quantized ConvTranspose"

Commit:
Update on "Base class for the quantized ConvTranspose"


Differential Revision: [D20641812](https://our.internmc.facebook.com/intern/diff/D20641812)

jerryzh168 commented on torch/nn/quantized/modules/conv.py:
```diff
@@ -12,13 +12,15 @@
-                 padding=0, dilation=1, groups=1, bias=True,
+                 padding=0, dilation=1,
+                 transposed=False, output_padding=0,
+                 groups=1, bias=True,
```
can we align with `_ConvNd` in `nn`? https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/conv.py#L20-L22, the default arguments seems to be put in the actual Conv modules like `Conv1d`, `Conv2d` etc.

Commit:
Update on "Base class for the quantized ConvTranspose"


Differential Revision: [D20641812](https://our.internmc.facebook.com/intern/diff/D20641812)

Commit:
Update on "Base class for the quantized ConvTranspose"


Differential Revision: [D20641812](https://our.internmc.facebook.com/intern/diff/D20641812)

jerryzh168 commented on torch/nn/quantized/modules/conv.py:
```diff
@@ -56,6 +58,8 @@ def extra_repr(self):
             s += ', padding={padding}'
         if self.dilation != (1,) * len(self.dilation):
             s += ', dilation={dilation}'
+        if self.output_padding != (0,) * len(self.output_padding):
```
what about `transposed`?

z-a-f replied:
I don't think we need it in the `str` because the name of the module already has transpose in it.

jerryzh168 commented on torch/nn/quantized/modules/conv.py:
```diff
@@ -498,3 +502,64 @@ def from_float(cls, mod):
+            res.append(pad)
+        return res
+
+    def _output_padding(self, input, output_size, stride, padding, kernel_size):
```
is this the same as: https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/conv.py#L496, can we reuse this code?

z-a-f replied:
Yes, I can inherit from the non-quantized class.

jerryzh168 reviewed (commented):
had a few comments

jerryzh168 commented on torch/nn/quantized/modules/conv.py:
```diff
@@ -498,3 +502,64 @@ def from_float(cls, mod):
+
+class _ConvTransposeNd(_ConvNd):
+    def __init__(self, in_channels, out_channels, kernel_size, stride,
+                 padding, dilation, transposed, output_padding,
```
also I think we probably don't need transposed, it should always be true for ConvTranspose right?

z-a-f replied:
This is forr the consistency with the FP codebase

jerryzh168 replied:
this doesn't make sense, shall we change it?

jerryzh168 replied:
can be a different PR

Commit:
Update on "Base class for the quantized ConvTranspose"


Differential Revision: [D20641812](https://our.internmc.facebook.com/intern/diff/D20641812)

Commit:
Update on "Base class for the quantized ConvTranspose"


Differential Revision: [D20641812](https://our.internmc.facebook.com/intern/diff/D20641812)

Commit:
Update on "Base class for the quantized ConvTranspose"


Differential Revision: [D20641812](https://our.internmc.facebook.com/intern/diff/D20641812)

jerryzh168 reviewed (approved):
Looks good, shall we remove `transpose` argument in `_ConvTransposeNd`?

z-a-f commented:
> Looks good, shall we remove `transpose` argument in `_ConvTransposeNd`?

We could, but in that case it will be inconsistent with the `nn.modules.conv._ConvTransposeNd`. Do you think we should still do it?

jerryzh168 commented:
> > Looks good, shall we remove `transpose` argument in `_ConvTransposeNd`?
> 
> We could, but in that case it will be inconsistent with the `nn.modules.conv._ConvTransposeNd`. Do you think we should still do it?

I think so, but you can have a separate PR

Commit:
Update on "Base class for the quantized ConvTranspose"


Differential Revision: [D20641812](https://our.internmc.facebook.com/intern/diff/D20641812)
